sdk/clientcore/http-netty4/src/main/java/io/clientcore/http/netty4/implementation/Netty4ChannelInputStream.java

```diff
@@ -0,0 +1,225 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+package io.clientcore.http.netty4.implementation;
+
+import io.netty.channel.Channel;
+
+import java.io.ByteArrayOutputStream;
+import java.io.InputStream;
+import java.util.LinkedList;
+import java.util.concurrent.CountDownLatch;
+
+/**
+ * Implementation of {@link InputStream} that reads contents from a Netty {@link Channel}.
+ */
+public final class Netty4ChannelInputStream extends InputStream {
+    private final Channel channel;
+
+    // Indicator for the Channel being fully read.
+    // This will become true before 'streamDone' becomes true, but both may become true in the same operation.
+    // Once this is true, the Channel will never be read again.
+    private boolean channelDone = false;
+
+    // Indicator for the stream being fully read.
+    // This will become true after 'channelDone' becomes true, but both may become true in the same operation.
+    // Once this is true, the stream will never return data again.
+    private boolean streamDone = false;
+
+    // Linked list of byte[]s that maintains the last available contents from the Channel / eager content.
+    // A list is needed as each Channel.read() may result in many channelRead calls.
+    private final LinkedList<byte[]> additionalBuffers;
+
+    private byte[] currentBuffer;
+
+    // Read index of the 'currentBuffer' that is actively being read.
+    // Once 'readIndex' == 'currentBuffer.length' 'additionalBuffers' will be polled for the next 'currentBuffer'.
+    // If 'additionalBuffers' doesn't contain anymore byte[]s 'readMore' will be called to attempt reading more data
+    // from the Channel. From there either more data will be read or 'streamDone = true' will be reached.
+    private int readIndex;
+
+    Netty4ChannelInputStream(ByteArrayOutputStream eagerContent, Channel channel) {
```

Comment:
can we have a doc on what exactly is `eagerContent`, it seems we're reading from this stream and channel

Reply:
I'll add documentation for this.

Effectively, when Netty reads from the `Channel` it may produce multiple `channelRead` calls, especially for initial response data. This is where we may get what I've called "eager" content where we need to handle it now but want to process it later on.

Reply:
Got it, thanks for the clarification